### PR TITLE
Implement fsevents

### PR DESCRIPTION
**Pull request by gsoltis**
Still WIP, but testing w/ `FSEvents` now works. 

Remaining:
  - [x] Clean up naming and package structure. It's possible we want to break out a `backend` package or something
  - [x] Verify behavior on remaining platforms (should be unchanged, but needs verification)
  - [ ] Documentation on code structure

### CODE REVIEW

File: cli/Makefile [hunk 1]
```diff
@@ -15,7 +15,7 @@ SRC_FILES = $(shell find . -name "*.go" | grep -v "_test.go")
 GENERATED_FILES = internal/turbodprotocol/turbod.pb.go internal/turbodprotocol/turbod_grpc.pb.go
 
 turbo: $(GENERATED_FILES) $(SRC_FILES) go.mod
-	CGO_ENABLED=0 go build $(GO_FLAGS) ./cmd/turbo
+	CGO_ENABLED=1 go build $(GO_FLAGS) ./cmd/turbo
```
Comment (nathanhammond) [inline cli/Makefile:18]: We need to make this platform-dependent.

**Review by nathanhammond — COMMENTED** [review]
Part one: the stuff that isn't `filewatcher`.

File: cli/internal/filewatcher/filewatcher_test.go [hunk 2]
```diff
@@ -76,13 +58,29 @@ func expectNoFilesystemEvent(t *testing.T, ch <-chan struct{}) {
 		} else {
 			t.Error("filewatching closed unexpectedly")
 		}
-	case <-time.After(100 * time.Millisecond):
+	case <-time.After(500 * time.Millisecond):
```
Comment (nathanhammond) [inline cli/internal/filewatcher/filewatcher_test.go:61]: Glad that this is single-invocation.

File: cli/internal/filewatcher/filewatcher_test.go [hunk 3]
```diff
@@ -102,64 +100,54 @@ func TestFileWatching(t *testing.T) {
 	//     child/
 	//     sibling/
 
-	watcher, err := fsnotify.NewWatcher()
+	watcher, err := GetPlatformSpecificBackend(logger)
 	assert.NilError(t, err, "NewWatcher")
```
Comment (nathanhammond) [inline cli/internal/filewatcher/filewatcher_test.go:104]: Some renaming in this area.

File: cli/internal/filewatcher/filewatcher_test.go [hunk 4]
```diff
@@ -102,64 +100,54 @@ func TestFileWatching(t *testing.T) {
...
-	expectFilesystemEvent(t, ch)
-	expectedEvent := fsnotify.Event{
-		Op:   fsnotify.Create,
-		Name: fooPath.ToString(),
-	}
-	c.mu.Lock()
-	got := c.createEvents[len(c.createEvents)-1]
-	c.mu.Unlock()
-	assert.DeepEqual(t, got, expectedEvent)
-	// Windows doesn't watch individual files, only directories
```
Comment (nathanhammond) [inline cli/internal/filewatcher/filewatcher_test.go:139]: This is a nice smoothing-over of the API.

File: cli/internal/filewatcher/backend.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+	"github.com/vercel/turborepo/cli/internal/doublestar"
+	"github.com/vercel/turborepo/cli/internal/fs"
+)
+
+// watchAddMode is used to indicate whether watchRecursively should synthesize events
+// for existing files.
+type watchAddMode int
+
+const (
+	dontSynthesizeEvents watchAddMode = iota
```
Comment (nathanhammond) [inline cli/internal/filewatcher/backend.go:25]: Not going to `iota + 1` since the values here map well to true/false.

**Review by nathanhammond — COMMENTED** [review]
All but the backends.

**Review by tknickman — APPROVED** [review]

**Pull request merged**